Implement scrollbar-color CSS parsing

lukewarlow opened the pull request:
#### 581332fbe81687fdac5e494e55c5364f99d7e7d6
<pre>
Implement scrollbar-color CSS parsing
<a href="https://bugs.webkit.org/show_bug.cgi?id=257571">https://bugs.webkit.org/show_bug.cgi?id=257571</a>

Reviewed by Darin Adler and Tim Nguyen.

Spec: <a href="https://drafts.csswg.org/css-scrollbars/#scrollbar-color">https://drafts.csswg.org/css-scrollbars/#scrollbar-color</a>

Grammar for property: auto | &lt;color&gt;{2}

This adds the property to the CSS parser behind a flag.
This also adds animation support for this property.
Test expectations have been updated, and new tests have been added.

* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/inheritance-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/property-list.js:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/property-types.js:
(const.colorPairType.testInterpolation):
(const.colorPairType.testAdditionOrAccumulation):
(const.colorPairType.testAddition):
(const.colorPairType.testAccumulation):
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::initialValueForLonghand):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeScrollSnapType):
(WebCore::CSSPropertyParserHelpers::consumeScrollbarColor):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialScrollbarColor):
(WebCore::RenderStyle::scrollbarColor const):
(WebCore::RenderStyle::scrollbarThumbColor const):
(WebCore::RenderStyle::scrollbarTrackColor const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setScrollbarColor):
(WebCore::RenderStyle::setScrollbarThumbColor):
(WebCore::RenderStyle::setScrollbarTrackColor):
* Source/WebCore/rendering/style/ScrollbarColor.cpp: Added.
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/ScrollbarColor.h: Added.
(WebCore::ScrollbarColor::MarkableTraits::isEmptyValue):
(WebCore::ScrollbarColor::MarkableTraits::emptyValue):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertScrollbarColor):

Canonical link: <a href="https://commits.webkit.org/265313@main">https://commits.webkit.org/265313@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4e2b66fc1322f630eedbfa16e0b651714bf7d797

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10586 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10808 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11087 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12232 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10157 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10600 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13179 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10773 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13086 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10747 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11673 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8902 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12633 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8967 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9560 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16821 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/8953 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10038 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9710 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12964 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/10043 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10176 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8256 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/10706 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9330 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/2862 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2530 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13594 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/11005 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10032 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2697 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -4191,6 +4191,11 @@ RefPtr<CSSValue> ComputedStyleExtractor::valueForPropertyInStyle(const RenderSty
+    case CSSPropertyScrollbarColor:
+        // TODO correctly handle inheritance
+        if (!style.scrollbarColor())
+            return CSSPrimitiveValue::create(CSSValueAuto);
```
Need to work out how to correctly handle the case where this is something like unset where instead of auto it should inherit. I believe the test failure is as a result of this line of code saying theres no value so it must be auto (which is correct) but unset should be handled. Now this should be handled I'm not yet sure.

Perhaps std::optional isn't the best data type? The equivalent is used inside of Chrome and that inherits fine though.

nt1m replied:
I think the issue is in CSSProperties.json, this code is fine

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -9303,6 +9303,20 @@
                 "url": "https://drafts.csswg.org/css-scroll-snap-1/#propdef-scroll-snap-stop"
             }
         },
+        "scrollbar-color": {
```
You're missing `"inherited": true,` here, hence the failure.

nt1m commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -1050,6 +1052,22 @@ inline ScrollSnapStop BuilderConverter::convertScrollSnapStop(BuilderState&, con
+{
+    if (is<CSSPrimitiveValue>(value)) {
+        auto& primitiveValue = downcast<CSSPrimitiveValue>(value);
+        if (primitiveValue.valueID() == CSSValueAuto)
```
You can ASSERT the valueID is auto, the parser should not pass bogus values

nt1m commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -435,6 +436,7 @@ inline GapLength RenderStyle::initialRowGap() { return { }; }
 constexpr RubyPosition RenderStyle::initialRubyPosition() { return RubyPosition::Before; }
 inline Length RenderStyle::initialScrollMargin() { return zeroLength(); }
 inline Length RenderStyle::initialScrollPadding() { return { }; }
+inline std::optional<class ScrollbarColor> RenderStyle::initialScrollbarColor() { return std::nullopt; }
```
`optional<class ScrollbarColor>` is a unusual choice. Usually we do something like `optional< ScrollbarColor>` or such.

lukewarlow replied:
I may be misunderstanding the issue but I repeatedly got errors about template arguments needing to be a type when I didn't specify class. Potentially due to scrollbar color being an enum value (the CSS keyword colour)

mdubet replied:
`ScrollbarColor` is not an enum ; `std::optional<ScrollbarColor>` should work fine.

lukewarlow replied:
ScrollbarColor is a value in the ThemeColor enum and that causes issues (fails to build) with referencing ScrollbarColor without the class keyword in front.

mdubet replied:
Indeed, the ThemeColor should be an `enum class` to prevent this.
Is there a way to avoid the include of `ThemeColor` though ?

mdubet replied:
https://github.com/WebKit/WebKit/pull/14730

lukewarlow replied:
Thanks for that. I'll rebase this and drop the class word, once that's merged

lukewarlow replied:
Made that and the other requested changes 🙂

mdubet commented on Source/WebCore/rendering/style/RenderStyleSetters.h:
```diff
@@ -508,6 +508,11 @@ inline void RenderStyle::setOrphans(unsigned short count)
     SET_PAIR(m_rareInheritedData, orphans, clampedCount, hasAutoOrphans, false);
 }
 
+inline void RenderStyle::setScrollbarColor(const std::optional<class ScrollbarColor> color)
```
Passing by const copy is unnecessary, we should pass it by const reference (or move it)

mdubet commented on Source/WebCore/rendering/style/ScrollbarColor.h:
```diff
@@ -0,0 +1,51 @@
+public:
+    ScrollbarColor(const StyleColor& thumbColor, const StyleColor& trackColor);
+
+    bool operator==(const ScrollbarColor& o) const
```
I think we can use the `= default` comparison operator now

mdubet commented on Source/WebCore/rendering/style/StyleRareInheritedData.h:
```diff
@@ -209,6 +210,8 @@ class StyleRareInheritedData : public RefCounted<StyleRareInheritedData> {
 
     ListStyleType listStyleType;
 
+    std::optional<class ScrollbarColor> scrollbarColor;
```
We should check with the `dump-class-layout`, but I think the `optional` will make it align at 12 bytes (4 for each StyleColor).
Using a custom `Markable<ScrollbarColor>` when any of the underlying color is invalid could save 4 bytes (might be worth writing it as a commit)

lukewarlow replied:
That command seg faults when I run it on my machine.

Apologies if this is a silly question, but there's only two StyleColours so where would the other 4 bytes be used I assume for the class instance itself?

Also this property cannot exist in a state with an invalid style colour. Either both the track and thumb colour are valid or this will be std::nullopt representing the auto value.

mdubet replied:
Optional takes 1 bit (and aligned on 4 bytes) which could be saved by using Markable (instead of having the `nullopt`, we would use the already existing invalid states of either of the StyleColor).

A comment should be enough, it could definitely be in a follow-up patch.

lukewarlow replied:
I will add a comment above this line that says

`// Investigate using Markable here to save on memory`

I've also made a follow up bug https://bugs.webkit.org/show_bug.cgi?id=257788

mdubet commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -1050,6 +1052,21 @@ inline ScrollSnapStop BuilderConverter::convertScrollSnapStop(BuilderState&, con
+    ASSERT(value.isPair());
+    const auto& pair = downcast<CSSValuePair>(value);
+
+    std::optional<class ScrollbarColor> result = *(new class ScrollbarColor(builderState.colorFromPrimitiveValueWithResolvedCurrentColor(downcast<CSSPrimitiveValue>(pair.first())), builderState.colorFromPrimitiveValueWithResolvedCurrentColor(downcast<CSSPrimitiveValue>(pair.second()))));
```
Why is it using `new` and they call copy actor ? It should just be a local stack value.

mdubet commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -967,6 +967,7 @@ class RenderStyle {
     const ScrollSnapAlign& scrollSnapAlign() const;
     ScrollSnapStop scrollSnapStop() const;
 
+    inline const std::optional<ScrollbarColor> scrollbarColor() const;
```
Do we need the `inline` here 🤔 ?

lukewarlow replied:
I assume yes because it's defined in the RenderStyleInlines.h file, but I may have misunderstood the connection between inline files and inline keyword.

lukewarlow replied:
If your question is related to whether it should be in RenderStyleInlines or RenderStyle then I'm also not sure on the decision that goes into choosing between them. It was rather arbitrary tbh. Any guidance on this would be great.

Ahmad-S792 replied:
I think @darinadler did changes around RenderStyle to speed-up builds and had separate 'RenderStyleInlines' as a result. I would personally just go back to particular commit and read commit message to understand the rationale, hope it answers. Just wanted to help how I can.

darinadler replied:
We use `inline` or `constexpr` when there is an inline function implementation in any other header. It guarantees we get a warning if we forget to include that header, even in builds with inlining disabled, like debug builds.

We don't need to use `inline` when the function body is in the same header. No harm, but we typically omit it.

@jernoble started doing this and I followed suit.

lukewarlow replied:
I've looked further into this but can't find a justification for when something should be inline vs just in the standard RenderStyle.cpp file.

I'm happy to leave it as inline or to move it if we don't think inlining is correct behaviour.

mdubet replied:
I think it's fine like this.
In general in the WebKit codebase, we keep bare getter/setter in headers and anything else in the implementation file.

mdubet commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -632,6 +634,7 @@ inline RotateTransformOperation* RenderStyle::rotate() const { return m_nonInher
 inline const GapLength& RenderStyle::rowGap() const { return m_nonInheritedData->rareData->rowGap; }
 inline RubyPosition RenderStyle::rubyPosition() const { return static_cast<RubyPosition>(m_rareInheritedData->rubyPosition); }
 inline ScaleTransformOperation* RenderStyle::scale() const { return m_nonInheritedData->rareData->scale.get(); }
+inline const std::optional<ScrollbarColor> RenderStyle::scrollbarColor() const { return m_rareInheritedData->scrollbarColor; }
```
Returning a `const` value doesn't make sense, we should either return a value or a const ref (a bit better for performance).

mdubet commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -1050,6 +1052,19 @@ inline ScrollSnapStop BuilderConverter::convertScrollSnapStop(BuilderState&, con
+    ASSERT(value.isPair());
+    const auto& pair = downcast<CSSValuePair>(value);
+
+    return std::make_optional<ScrollbarColor>(builderState.colorFromPrimitiveValueWithResolvedCurrentColor(downcast<CSSPrimitiveValue>(pair.first())), builderState.colorFromPrimitiveValueWithResolvedCurrentColor(downcast<CSSPrimitiveValue>(pair.second())));
```
We should not resolve currentColor here, we should use `BuilderState::colorFromPrimitiveValue`

nt1m reviewed (approved):
Looks good, but can you implement animation support?

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/property-list.js:
```diff
@@ -1172,6 +1172,10 @@ const gCSSProperties2 = {
       { type: 'discrete', options: [ [ 'auto', 'smooth' ] ] }
     ]
   },
+  'scrollbar-color': {
```
Is this an import of new changes from WPT or something you are doing in WebKit tree and plan to export? The change log only mentions updating expectation files, so does not make that clear.

lukewarlow replied:
It's a change in WebKit that I've exported to a WPT pr. Apologies I forgot to update that part of the commit message.

darinadler commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -1793,6 +1793,61 @@ class PropertyWrapperColor : public AnimationPropertyWrapperBase {
 };
 
+// TODO: How to deal with animating between scrollbar-color: auto and scrollbar-color: <color>{2}
+class ScrollbarColorPropertyWrapper : public AnimationPropertyWrapperBase {
```
Class could be `final`.

darinadler commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -1793,6 +1793,61 @@ class PropertyWrapperColor : public AnimationPropertyWrapperBase {
+    }
+
+protected:
+    bool equals(const RenderStyle& a, const RenderStyle& b) const override
```
Could be `final` rather than `override`.

darinadler commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -1793,6 +1793,61 @@ class PropertyWrapperColor : public AnimationPropertyWrapperBase {
+
+    bool canInterpolate(const RenderStyle& from, const RenderStyle& to, CompositeOperation) const final
+    {
+        return from.scrollbarColor().has_value() && to.scrollbarColor().has_value();
```
Should we avoid wastefully creating the wrappers in the case where `canInterpolate` is false?

lukewarlow replied:
I'm not sure how to go about doing this?

darinadler replied:
I think we can’t.

darinadler commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -1793,6 +1793,61 @@ class PropertyWrapperColor : public AnimationPropertyWrapperBase {
+        return from.scrollbarColor().has_value() && to.scrollbarColor().has_value();
+    }
+
+    void blend(RenderStyle& destination, const RenderStyle& from, const RenderStyle& to, const CSSPropertyBlendingContext& context) const override
```
Could be `final` rather than `override`.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -6671,6 +6671,19 @@ RefPtr<CSSValue> consumeScrollSnapType(CSSParserTokenRange& range)
+RefPtr<CSSValue> consumeScrollbarColor(CSSParserTokenRange& range, const CSSParserContext& context)
+{
+    if (auto ident = consumeIdent<CSSValueAuto>(range))
+        return CSSPrimitiveValue::create(CSSValueAuto);
```
Should be:
```
    return ident;
```

darinadler commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -1017,4 +1020,18 @@ inline bool pseudoElementRendererIsNeeded(const RenderStyle* style)
 
+inline const StyleColor& RenderStyle::scrollbarThumbColor() const
+{
+    ASSERT(m_rareInheritedData->scrollbarColor.has_value());
```
This assertion is not necessary and should be omitted. The `std::optional` class already takes care of checking for null and stopping the program and we need not also assert an optional is non-null every time we dereference it.

Removing this turns this into a one-liner that should be in the block of one-liners above.

darinadler commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -1017,4 +1020,18 @@ inline bool pseudoElementRendererIsNeeded(const RenderStyle* style)
+
+inline const StyleColor& RenderStyle::scrollbarTrackColor() const
+{
+    ASSERT(m_rareInheritedData->scrollbarColor.has_value());
```
Same here.

darinadler commented on Source/WebCore/rendering/style/RenderStyleSetters.h:
```diff
@@ -503,6 +503,25 @@ inline void RenderStyle::setOrphans(unsigned short count)
     SET_PAIR(m_rareInheritedData, orphans, clampedCount, hasAutoOrphans, false);
 }
 
+inline void RenderStyle::setScrollbarColor(const std::optional<ScrollbarColor>& color)
```
This is a one-liner that should be in the block of one-liners above.

darinadler commented on Source/WebCore/rendering/style/RenderStyleSetters.h:
```diff
@@ -503,6 +503,25 @@ inline void RenderStyle::setOrphans(unsigned short count)
+
+inline void RenderStyle::setScrollbarThumbColor(const StyleColor& color)
+{
+    ASSERT(m_rareInheritedData->scrollbarColor.has_value());
```
This is another unnecessary `std::optional` assertion.

darinadler commented on Source/WebCore/rendering/style/RenderStyleSetters.h:
```diff
@@ -503,6 +503,25 @@ inline void RenderStyle::setOrphans(unsigned short count)
+
+inline void RenderStyle::setScrollbarTrackColor(const StyleColor& color)
+{
+    ASSERT(m_rareInheritedData->scrollbarColor.has_value());
```
This is another unnecessary `std::optional` assertion.

darinadler commented on Source/WebCore/rendering/style/ScrollbarColor.cpp:
```diff
@@ -0,0 +1,45 @@
+    ts << scrollbarColor.thumbColor();
+    ts << " ";
+    ts << scrollbarColor.trackColor();
+    return ts;
```
Should be:
```
    return ts << scrollbarColor.thumbColor() << ' ' << scrollbarColor.trackColor();
```
Or similar.

darinadler commented on Source/WebCore/rendering/style/ScrollbarColor.h:
```diff
@@ -0,0 +1,58 @@
+
+namespace WebCore {
+
+class ScrollbarColor {
```
This would be better as a `struct`. There is no good reason to have getters and setters. It’s just a data holder that holds two colors:
```
struct ScrollbarColors {
    StyleColor thumbColor;
    StyleColor trackColor;
};

bool operator==(const ScrollbarColors&, const ScrollbarColors&) = default;
WTF::TextStream& operator<<(WTF::TextStream&, const ScrollbarColors&);
```

darinadler commented on Source/WebCore/rendering/style/StyleRareInheritedData.h:
```diff
@@ -204,6 +205,9 @@ class StyleRareInheritedData : public RefCounted<StyleRareInheritedData> {
 
     WordBoundaryDetection wordBoundaryDetection;
 
+    // Investigate using Markable here to save on memory
```
This would be very easy to do since `StyleColor` has a null value, and I don’t see why we wouldn’t do it right away instead of just leaving a comment.

lukewarlow replied:
I'm not sure how to go about doing this? When I tried adding a MarkableTraits struct to the ScrollbarColor struct I got errors such as `Incomplete type 'WebCore::ScrollbarColor' named in nested name specifier` I believe this is something to do with how the ScrollbarColor type is forwarded inside of RenderStyle.h. I had other related errors too

(I'm aware the below code is almost definitely not fully correct)

```cpp
struct MarkableTraits {
        static bool isEmptyValue(const ScrollbarColor& value)
        {
            return value.thumbColor == StyleColor { Color(std::nullopt) } && value.trackColor == StyleColor(Color(std::nullopt));
        }

        static ScrollbarColor emptyValue()
        {
            return ScrollbarColor {
                StyleColor { Color(std::nullopt) },
                StyleColor { Color(std::nullopt) }
            };
        }
    };
```

lukewarlow replied:
I'll push up the other requested changes such as changing to a struct and the code cleanup. And then hopefully someone can take a look at this to point me in the right direction.

darinadler replied:
There are really two separate problems:
1) How to declare the markable traits so they will be used correctly in `RenderStyle.h`. One solution there is to have the functions return `std::optional<ScrollbarColor>` instead of `const Markable< ScrollbarColor>&`, which probably will work fine from a performance point of view and doesn’t require any markable traits in that header at all. It’s worth looking for another example of use of `Markable` in `RenderStyle` for a view on what idiom they use.
2) How to define the markable traits in the `ScrollbarColor.h` header so they work correctly for **use** of `Markable<StyleColor>`.

It should be closer to this (but this is not exactly right):
```
    template<> struct MarkableTraits<StyleColor> {
        static bool isEmptyValue(const ScrollbarColors& value) { return !value.thumbColor.isValid(); };
        static ScrollbarColors emptyValue() { return { } };
    }
```
Might need to add some `{ }` to the definition of the ScrollbarColors struct to make this work.

lukewarlow replied:
I couldn't find an example of Markable in RenderStyle but I *think* I've worked it out. The tests seemed to pass locally so will see what the pipeline thinks of it.

darinadler commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -44,6 +44,7 @@
 #include "RenderStyleSetters.h"
 #include "RenderTheme.h"
 #include "ScaleTransformOperation.h"
+#include "ScrollbarColor.h"
```
This include is not needed. Please don’t add it.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/property-types.js:
It does seem like we're not testing animating pairs of mismatching colors (between track & thumb), which we should test, since it's the most common case with scrollbar-color

lukewarlow replied:
Ah yeah that makes sense, oversight on my part too focussed on getting the basic animation working.

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -9018,6 +9018,24 @@
+            },
+            "status": {
+                "status": "in development"
+            }
```
```suggestion
            "status": "in development"
```

nt1m commented on Source/WebCore/rendering/style/ScrollbarColor.h:
```diff
@@ -0,0 +1,56 @@
+    struct MarkableTraits {
+        static bool isEmptyValue(const ScrollbarColor& value)
+        {
+            return !value.thumbColor.absoluteColor().isValid();
```
This is likely going to break if you `color-mix` with `currentColor` or just `currentColor` by itself.

lukewarlow replied:
What's the correct way to access the isValid? There doesn't seem to be one directly on StyleColor? Do I need to resolve instead?

darinadler replied:
Oh, we need to initialize these `StyleColor` to invalid colors (which can be done with `Color { }`), not `currentColor`. I had no idea that the default constructor for `StyleColor` creates `currentColor`!

nt1m replied:
`StyleColor::absoluteColor()` will assert when called with something not absolute (e.g. currentColor/color-mix()) though. Not sure what's a good check for this though.

darinadler replied:
I think it needs to be this:
```
    return value.thumbColor.isAbsoluteColor() && !value.thumbColor.absoluteColor().isValid()
```

darinadler commented on Source/WebCore/rendering/style/ScrollbarColor.h:
```diff
@@ -0,0 +1,56 @@
+namespace WebCore {
+
+struct ScrollbarColor {
+    StyleColor thumbColor { Color(std::nullopt) };
```
This should be:
```
    StyleColor thumbColor { Color { } };
```

darinadler commented on Source/WebCore/rendering/style/ScrollbarColor.h:
```diff
@@ -0,0 +1,56 @@
+
+namespace WebCore {
+
+struct ScrollbarColor {
```
I suggest this be named `ScrollbarColors`, plural.

lukewarlow replied:
I went with ScrollbarColor to match the CSS Property but happy to change it?

darinadler replied:
Doesn’t seem critical either way. I can think of more arguments but nothing compelling.

darinadler commented on Source/WebCore/rendering/style/ScrollbarColor.h:
```diff
@@ -0,0 +1,56 @@
+
+struct ScrollbarColor {
+    StyleColor thumbColor { Color(std::nullopt) };
+    StyleColor trackColor { Color(std::nullopt) };
```
```
    StyleColor trackColor { Color { } };
```

darinadler commented on Source/WebCore/rendering/style/ScrollbarColor.h:
```diff
@@ -0,0 +1,56 @@
+
+        static ScrollbarColor emptyValue()
+        {
+            return { };
```
This needs to be:
```
    return { Color { }, Color { } };
```
So we get invalid color instead of `currentColor` for `thumbColor`, which is what makes this an empty value.

lukewarlow replied:
Would { } not already use the default values assigned inside the struct (both being assigned Color { })?

darinadler replied:
It would.

But note that typically we would not want the default value for a `ScrollbarColor` to be the one that will confuse `Markable` instead there would be one value that is the default for general use and another that is the value used for "null" in `Markable`.

darinadler commented on Source/WebCore/rendering/style/ScrollbarColor.h:
```diff
@@ -0,0 +1,56 @@
+    StyleColor trackColor { Color(std::nullopt) };
+
+    ScrollbarColor() { };
+    ScrollbarColor(const StyleColor& thumbColor, const StyleColor& trackColor);
```
Please remove both of these.

darinadler replied:
We don’t need a constructor to initialize `struct` members. We might have to change the syntax at the call site to use `{ }` instead of `( )`, but no need for a constructor since we can use aggregate initialization.

darinadler commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -1054,6 +1056,24 @@ inline ScrollSnapStop BuilderConverter::convertScrollSnapStop(BuilderState&, con
+        builderState.colorFromPrimitiveValue(downcast<CSSPrimitiveValue>(pair.second()))
+    };
+
+    return std::make_optional<ScrollbarColor>(result);
```
This will copy the colors, which is wasteful if they are complex colors. Easy to avoid. Here is one way.
```
return WTFMove(result);
```
Another way:
```
return ScrollbarColor {
    builderState.colorFromPrimitiveValue(downcast<CSSPrimitiveValue>(pair.first())),
    builderState.colorFromPrimitiveValue(downcast<CSSPrimitiveValue>(pair.second()))
};
```

lukewarlow replied:
Okay thanks!

darinadler commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -1054,6 +1056,24 @@ inline ScrollSnapStop BuilderConverter::convertScrollSnapStop(BuilderState&, con
+        return std::nullopt;
+    }
+
+    ASSERT(value.isPair());
```
This assertion is the same assertion that `downcast<>` does, so it should not be coded separately and redundantly.

darinadler commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -1054,6 +1056,24 @@ inline ScrollSnapStop BuilderConverter::convertScrollSnapStop(BuilderState&, con
+    }
+
+    ASSERT(value.isPair());
+    const auto& pair = downcast<CSSValuePair>(value);
```
Plain old `auto&` would be just fine. The `const` here doesn't really do anything.

darinadler commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -1793,6 +1793,60 @@ class PropertyWrapperColor : public AnimationPropertyWrapperBase {
+        destination.setScrollbarColor(blendingRenderStyle.scrollbarColor());
+    }
+
+    std::unique_ptr<PropertyWrapperStyleColor> m_thumbWrapper;
```
Something I have not considered before and I will look into is why we use `unique_ptr` at all for property wrappers instead of embedding objects.

darinadler commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -1793,6 +1793,60 @@ class PropertyWrapperColor : public AnimationPropertyWrapperBase {
+    {
+    }
+
+protected:
```
A `final` class does not need anything `protected`. Should just be `private`. Not important though.

darinadler commented on Source/WebCore/rendering/style/ScrollbarColor.h:
```diff
@@ -0,0 +1,53 @@
+
+        static ScrollbarColor emptyValue()
+        {
+            return { Color { }, Color { } };
```
As you pointed out, if we do make the colors default to invalid, as we do in this patch, then we can just use `{ }` here.

I think the best option might be to just omit the member initial values above and let them default to default construction, a little cleaner I think.

lukewarlow committed:
Implement scrollbar-color CSS parsing
https://bugs.webkit.org/show_bug.cgi?id=257571

Reviewed by Darin Adler and Tim Nguyen.

Spec: https://drafts.csswg.org/css-scrollbars/#scrollbar-color

Grammar for property: auto | <color>{2}

This adds the property to the CSS parser behind a flag.
This also adds animation support for this property.
Test expectations have been updated, and new tests have been added.

* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/inheritance-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/property-list.js:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/property-types.js:
(const.colorPairType.testInterpolation):
(const.colorPairType.testAdditionOrAccumulation):
(const.colorPairType.testAddition):
(const.colorPairType.testAccumulation):
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::initialValueForLonghand):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeScrollSnapType):
(WebCore::CSSPropertyParserHelpers::consumeScrollbarColor):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialScrollbarColor):
(WebCore::RenderStyle::scrollbarColor const):
(WebCore::RenderStyle::scrollbarThumbColor const):
(WebCore::RenderStyle::scrollbarTrackColor const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setScrollbarColor):
(WebCore::RenderStyle::setScrollbarThumbColor):
(WebCore::RenderStyle::setScrollbarTrackColor):
* Source/WebCore/rendering/style/ScrollbarColor.cpp: Added.
(WebCore::operator<<):
* Source/WebCore/rendering/style/ScrollbarColor.h: Added.
(WebCore::ScrollbarColor::MarkableTraits::isEmptyValue):
(WebCore::ScrollbarColor::MarkableTraits::emptyValue):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertScrollbarColor):

Canonical link: https://commits.webkit.org/265313@main
